File: kernel/src/main/kotlin/ninja/blacknet/network/packet/GetBlocks.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.network.packet

import kotlinx.serialization.Serializable
import ninja.blacknet.Kernel
import ninja.blacknet.crypto.Hash
import ninja.blacknet.db.CoinDB
import ninja.blacknet.network.Connection
import ninja.blacknet.network.Node

@Serializable
class GetBlocks(
    private val best: Hash,
    private val checkpoint: Hash
) : Packet {
    override fun handle(connection: Connection) {
        val cachedBlock = Kernel.blockDB().cachedBlock
        if (cachedBlock != null) {
            val (previousHash, bytes) = cachedBlock
            if (previousHash == best) {
                connection.sendPacket(PacketType.Blocks, Blocks(emptyList(), listOf(bytes)))
                return
            }
        }

        var blockIndex = CoinDB.blockIndexes.get(best.bytes)

        if (blockIndex == null) {
            val nextBlockHashes = CoinDB.getNextBlockHashes(checkpoint, Blocks.MAX_HASHES)
            if (nextBlockHashes != null) {
                connection.sendPacket(PacketType.Blocks, Blocks(nextBlockHashes, emptyList()))
                return
            } else {
                connection.sendPacket(PacketType.ConsensusFault, ConsensusFault())
                connection.dos("Consensus fault")
                return
            }
        }

        var size = PACKET_HEADER_SIZE_BYTES + 2 + 1
        val maxSize = Node.getMinPacketSize() // we don't know actual value, so assume minimum
        val response = ArrayList<ByteArray>()

        while (true) {
            if (blockIndex == null)
                break
            val hash = blockIndex.next
            if (hash == Hash.ZERO)
                break
            size += blockIndex.nextSize + 4 //TODO VarInt.size()
            if (response.isNotEmpty() && size >= maxSize)
                break
            val bytes = Kernel.blockDB().blocks.getBytes(hash.bytes)
            if (bytes == null)
                break
            response.add(bytes)
            if (response.size == Blocks.MAX_BLOCKS)
                break
            blockIndex = CoinDB.blockIndexes.get(hash.bytes)
        }

        connection.sendPacket(PacketType.Blocks, Blocks(emptyList(), response))
    }
}

File: network/src/main/rust/packet/getblocks.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::connection::Connection;
use crate::packet::{
    Blocks, ConsensusFault, MAX_BLOCKS, MAX_HASHES, PACKET_HEADER_SIZE_BYTES, Packet,
};
use blacknet_kernel::blake2b::Hash;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
pub struct GetBlocks {
    best: Hash,
    checkpoint: Hash,
}

impl Packet for GetBlocks {
    fn handle(self, connection: &mut Connection) {
        let node = connection.node();
        let block_db = node.block_db();
        if let Some((previous_hash, bytes)) = &**block_db.cached_block()
            && self.best == *previous_hash
        {
            connection.send_packet(Blocks::with_block(bytes));
            return;
        }

        if let Some(mut block_index) = block_db.index(self.best) {
            let mut size = PACKET_HEADER_SIZE_BYTES + 2 + 1;
            let max_size = node.min_packet_size(); // actual value is unknown, minimum is assumed
            let mut response = Vec::<Box<[u8]>>::with_capacity(MAX_BLOCKS);

            loop {
                let hash = block_index.next();
                if hash == Hash::default() {
                    break;
                }
                size += block_index.next_size() + 4; //XXX VarInt.size()
                if !response.is_empty() && size >= max_size {
                    break;
                }
                if let Some(bytes) = block_db.get_raw(hash) {
                    response.push(bytes);
                } else {
                    break;
                }
                if response.len() == MAX_BLOCKS {
                    break;
                }
                if let Some(next_index) = block_db.index(hash) {
                    block_index = next_index;
                } else {
                    break;
                }
            }

            connection.send_packet(Blocks::with_blocks(response));
        } else if let Some(next_block_hashes) =
            block_db.next_block_hashes(self.checkpoint, MAX_HASHES)
        {
            connection.send_packet(Blocks::with_hashes(next_block_hashes));
        } else {
            connection.send_packet(ConsensusFault);
            connection.dos("Consensus fault");
        }
    }
}
